### PR TITLE
Status check

### PR DESCRIPTION
**Pull request by jsliacan**
**Fixes:** Issue #1737 

## Solution

The entire `crc status` check is hidden in one function. If looking for `Stopped` status, then exit with success as soon as status returns `Stopped` for the first time. If looking for `Running` status, then exit with success only if 3 consecutive probes (at 1min intervals) return `Running` cluster status. The probing is set to 15x1min.

Second commit updates the language in Feature files.

## Testing

1. `make integration` succeeds

### CODE REVIEW

File: test/integration/crcsuite/crcsuite.go [hunk 1]
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
...
+	retryCount := 15
+	cmd := "crc status --log-level debug"
+	expression := ``
+	var numConsecutive int
+
+	if state == "Stopped" {
+		numConsecutive = 1
+		expression = `.*Stopped.*`
+	} else {
+		numConsecutive = 3
```
Comment (praveenkumar) [inline test/integration/crcsuite/crcsuite.go:266]: So this number is present to have consecutive execute the crc status and see if have same output?

File: test/integration/crcsuite/crcsuite.go [hunk 2]
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
 	return matchErr
 }
 
+// CheckCRCStatus checks that output of status command
+// matches given regex number of consecutive times
+func CheckCRCStatus(state string) error {
+
```
Comment (praveenkumar) [inline test/integration/crcsuite/crcsuite.go:255]: unwanted new line

File: test/integration/crcsuite/crcsuite.go [hunk 3]
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
...
+			}
+			// break if done
+			if count == numConsecutive {
+				return nil
+			}
+		}
+		time.Sleep(retryDuration)
+	}
+
+	return fmt.Errorf("did not get enough consecutive matches")
```
Comment (praveenkumar) [inline test/integration/crcsuite/crcsuite.go:291]: We should also mention how much matches we wanted and how much is present.

File: test/integration/features/basic.feature [hunk 4]
```diff
@@ -21,18 +21,18 @@ Feature: Basic test
         And stdout should contain "delete"
         And stdout should contain "status"
         And stdout should contain "Flags:"
-        And stdout should contain 
+        And stdout should contain
```
Comment (praveenkumar) [inline test/integration/features/basic.feature:26]: This shouldn't be part of this commit.

File: test/integration/features/basic.feature [hunk 5]
```diff
@@ -21,18 +21,18 @@ Feature: Basic test
...
     Scenario: CRC status
         When executing "crc status" fails
-        Then stderr should contain 
-        """
-        Machine 'crc' does not exist. Use 'crc start' to create it
-        """
+        Then stderr should contain
+            """
+            Machine 'crc' does not exist. Use 'crc start' to create it
+            """
```
Comment (praveenkumar) [inline test/integration/features/basic.feature:37]: This shouldn't be part of this commit.

File: test/integration/features/cert_rotation.feature [hunk 6]
```diff
@@ -1,27 +1,27 @@
...
+    And executing "sudo timedatectl set-ntp on" succeeds
 
-    Scenario: CRC delete and cleanup
-      When executing "crc delete -f" succeeds
-      Then stdout should contain "Deleted the OpenShift cluster"
-      When executing "crc cleanup" succeeds
+  Scenario: CRC delete and cleanup
+    When executing "crc delete -f" succeeds
+    Then stdout should contain "Deleted the OpenShift cluster"
+    When executing "crc cleanup" succeeds
```
Comment (praveenkumar) [inline test/integration/features/cert_rotation.feature:27]: Formatting should be done is different commit message

File: test/integration/features/story_marketplace.feature [hunk 7]
```diff
@@ -68,7 +68,7 @@ Feature:
         When executing "oc apply -f etcd-cluster3.yaml" succeeds
         Then with up to "10" retries with wait period of "30s" command "oc get pods" output matches "(?s)(.*example-[a-z0-9]* *1/1 *Running.*){3}"
         But with up to "10" retries with wait period of "30s" command "oc get pods" output does not match "(?s)(.*example-[a-z0-9]* *1/1 *Running.*){4}"
-        
+
```
Comment (praveenkumar) [inline test/integration/features/story_marketplace.feature:72]: This shouldn't be part of this commit.

**Review by praveenkumar — CHANGES_REQUESTED** [review]

File: test/integration/crcsuite/crcsuite.go (hunk 1 above)
Comment (jsliacan) [reply test/integration/crcsuite/crcsuite.go:266]: Yes. Exactly. In particular, I assume that if OCP cluster has been Running for 3mins in a row, then it's "stable" for the rest of the Feature (in particular, this check avoids "Running -> Not reachable -> Starting" pitfall right after start.

File: test/integration/features/basic.feature (hunk 4 above)
Comment (jsliacan) [reply test/integration/features/basic.feature:26]: OK, will address all of these.

File: test/integration/features/proxy.feature [hunk 8]
```diff
@@ -1,9 +1,8 @@
 @proxy @linux
-Feature: Test the proxy
+Feature: Behind proxy test
 
-  The user tries to use crc behind proxy. he/she expects the
-  crc start happen successfully and able to deploy the app and check its
-  accessibility.
+  User starts CRC behind proxy. They expect a successful start
```
Comment (cfergeau) [inline test/integration/features/proxy.feature:4]: `behind a proxy` maybe?

File: test/integration/features/story_health.feature [hunk 9]
```diff
@@ -1,8 +1,10 @@
...
-Feature:
-    End-to-end health check. Set-up and start CRC. Then create a
-    project and deploy an app. Check on the app and delete the
-    project. Stop and delete CRC.
+Feature: End-to-end health check
+
+    User goes through a "day in the life of" CRC. They set-up
+    and start CRC. They then create a project and deploy an app.
+    They check on the app and delete the project. They stop and
+    delete CRC.
```
Comment (cfergeau) [inline test/integration/features/story_health.feature:7]: Imo it's weird to talk about "deleting CRC", it's the CRC VM or instance which is being acted on.

File: test/integration/features/story_marketplace.feature [hunk 10]
```diff
@@ -1,7 +1,8 @@
 @story_marketplace
-Feature:
-    Install OpenShift operator from OperatorHub and use it to manage
-    admin tasks.
+Feature: Operator from marketplace
+
+    User installs OpenShift operator from OperatorHub and uses it
```
Comment (cfergeau) [inline test/integration/features/story_marketplace.feature:4]: `an OpenShift operator` ?

File: test/integration/crcsuite/crcsuite.go [hunk 11]
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
 	return matchErr
 }
 
+// CheckCRCStatus checks that output of status command
+// matches given regex number of consecutive times
+func CheckCRCStatus(state string) error {
+
+	retryDuration, _ := time.ParseDuration("1m")
```
Comment (cfergeau) [inline test/integration/crcsuite/crcsuite.go:256]: I think it can just be `retryDuration := 1 * time.Minute`

File: test/integration/crcsuite/crcsuite.go [hunk 12]
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
...
+	retryCount := 15
+	cmd := "crc status --log-level debug"
+	expression := ``
```
Comment (cfergeau) [inline test/integration/crcsuite/crcsuite.go:259]: You can init with `""`

File: test/integration/crcsuite/crcsuite.go [hunk 13]
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
...
+	var numConsecutive int
+
+	if state == "Stopped" {
+		numConsecutive = 1
+		expression = `.*Stopped.*`
```
Comment (cfergeau) [inline test/integration/crcsuite/crcsuite.go:264]: Here you can also use `".*Stopped.*"` I think.

File: test/integration/crcsuite/crcsuite.go [hunk 14]
```diff
@@ -247,6 +249,48 @@ func CheckOutputMatchWithRetry(retryCount int, retryTime string, command string,
...
+	} else {
+		numConsecutive = 3
+		expression = `.*Running \(v\d+\.\d+\.\d+.*\).*`
+	}
```
Comment (cfergeau) [inline test/integration/crcsuite/crcsuite.go:268]: And here the backticks are useful to have because of the `\`

File: test/integration/crcsuite/crcsuite.go [hunk 15]
```diff
@@ -54,6 +54,8 @@ func FeatureContext(s *godog.Suite) {
 		CheckHTTPResponseWithRetry)
 	s.Step(`^with up to "(\d+)" retries with wait period of "(\d*(?:ms|s|m))" command "(.*)" output (should match|matches|should not match|does not match) "(.*)"$`,
 		CheckOutputMatchWithRetry)
+	s.Step(`^checking that CRC status is (Running|Stopped)$`,
```
Comment (cfergeau) [inline test/integration/crcsuite/crcsuite.go:57]: nicer to read as `checking that CRC is (running|stopped)$` ?

**Review by cfergeau — COMMENTED** [review]
Couple of comments, but feel free to ignore :)

File: test/integration/crcsuite/crcsuite.go (hunk 14 above)
Comment (jsliacan) [reply test/integration/crcsuite/crcsuite.go:268]: I used backticks everywhere because I had to use them here. Will change other places to double quotes.

File: test/integration/crcsuite/crcsuite.go (hunk 15 above)
Comment (jsliacan) [reply test/integration/crcsuite/crcsuite.go:57]: True. Will address other comments as well. Thanks!

File: test/integration/crcsuite/crcsuite.go (hunk 2 above)
Comment (jsliacan) [reply test/integration/crcsuite/crcsuite.go:255]: I'm not sure if it's unwanted (gofmt doesn't mind and the same newlines are in all other functions). But I don't mind, so I removed it. I will go through these files later and also give comments to all exported functions.

**Comment by jsliacan** [comment]
@praveenkumar @cfergeau , I addressed all comments (I hope!). Kept 1st commit as the main one. Second improves Feature descriptions and such, and I moved all formatting changes to a new (3rd) commit.

File: test/integration/features/basic.feature (hunk 4 above)
Comment (gbraad) [reply test/integration/features/basic.feature:26]: ETA?

**Review by gbraad — CHANGES_REQUESTED** [review]
whitespaces/indentations are all over the place

**Comment by jsliacan** [comment]
> whitespaces/indentations are all over the place

@gbraad , the purpose of the 3rd commit was to clean that up (when different people write Cucumber code, it indents with 2/4 spaces (even 3!), sometimes with tabs, etc.). Now the Features should be pretty consistent. Any examples of bad indentation?
